### ISSUE TITLE
AOT build silently ignores dynamic route configuration

### ISSUE THREAD

**Issue by peterbakonyi05** [angular/angular#20239]
## I'm submitting a...
<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior
When route configuration contains some shared static data, that data will be ignored during AOT compilation. However it works as expected with JIT compiler.
```ts
const sharedConfig = {
  data: {
    trackingArea: 'some-feature',
    product: 'some-feature'
  }
};

// `data` property will not be present with AOT
const routes: Routes = [
  { path: '', pathMatch: 'full', component: HomeComponent, ...sharedConfig },
  { path: '**', component: NotFoundComponent, ...sharedConfig }
];

@NgModule({
  imports: [RouterModule.forRoot(routes)],
  exports: [RouterModule]
})
export class AppRoutingModule { }
```

## Expected behavior
AOT build appends `data` to the route config (or at least fails with a meaningful error message).

## Minimal reproduction of the problem with instructions
https://github.com/peterbakonyi05/ngx.dynamic-route-config-bug

Start the attached angular cli application with AOT and with JIT build and compare the route config in the console. Check the README of the repo to see how to start the app (`npm install` and then `npm start` for JIT or `npm run start-prod` for AOT)

## What is the motivation / use case for changing the behavior?
Some of the routes needs the same static configuration in the `data` property. Without not sharing the config, the route configuration will be quite verbose and not DRY at all (~30% more config with a lot of duplication).

## Environment

<pre><code>
Angular version: 5.0.0

Browser:
- [x ] Chrome (desktop) version 61
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [ ] Firefox version XX
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX
 
For Tooling issues:
- Node version: 8.9.0
- Platform:  Windows 10

Others:
</code></pre>

**Comment by jpleclerc** [comment]
Any plans on fixing this issue?

**Comment by HelloQingGuo** [comment]
yes, bundling with AOT, dynamic routing  will fail silently with no error prompt in the console.
The picture below shows that I have a dynamically spawned routings `appRoutes` based on state name by using a simple for loop. The routing array was generated properly, but the it actually did not work (not injecting the component which matched the router).
So I have to remove the for loop, and hard code it ?

BTW, the interesting part from my experience is the for-loop is accepted in the routing when bundling with JIT (`ng serve`), while something like `Array.concat` and `spread operator` is not acceptable (it will throw error).

![image](https://user-images.githubusercontent.com/31576447/43966890-126689c2-9c91-11e8-91f4-6949471f29fb.png)

**Comment by benelliott** [comment]
This seems to be fixed as of v8.0.0

**Comment by HelloQingGuo** [comment]
Thanks for the updates!

**Comment by googlicius** [comment]
I also could not using `spread operator` and `Object.assign` in router resolve.

**Comment by lixaotec** [comment]
I´m having the same issue with angular 6.2

**Comment by JoostK** [comment]
A bit late, but good news: this has been fixed with Ivy in Angular 9.0!
